### PR TITLE
Fix issues

### PR DESCRIPTION
**Pull request by mfeurer**
Fixes jamming the server with too many concurrent requests by adding a new flag.

### CODE REVIEW

File: doc/progress.rst [hunk 1]
```diff
@@ -20,6 +20,8 @@ Changelog
   OpenML.
 * ADD #564: New helpers to access the structure of a flow (and find its
   subflows).
+* ADD #618: The software will from now on retry to connect to the server if a
+  connection failed. The number of retries can be configured.
```
Comment (janvanrijn) [inline doc/progress.rst]: can you also add #602 please? maybe close to columntransformer (sorry i forgot)

File: openml/_api_calls.py [hunk 2]
```diff
@@ -102,12 +103,39 @@ def _read_url(url, data=None):
     return response.text
 
 
+def send_request(request_method, url, kwargs):
```
Comment (janvanrijn) [inline openml/_api_calls.py]: i hate kwargs, it's so vague. Do we need this level of flexibility or can we just add optional arguments?

File: doc/progress.rst (hunk 1 above)
Comment (mfeurer) [reply doc/progress.rst]: As the column transformer is only added in this new release it makes no sense to also add a fix to this new features in the change log.

File: openml/_api_calls.py (hunk 2 above)
Comment (mfeurer) [reply openml/_api_calls.py]: Changed.

File: openml/_api_calls.py [hunk 3]
```diff
@@ -1,22 +1,17 @@
...
-import arff
 import xmltodict
 
 from . import config
 from .exceptions import (OpenMLServerError, OpenMLServerException,
                          OpenMLServerNoResult)
 
 
-def _perform_api_call(call, data=None, file_elements=None,
-                      add_authentication=True):
```
Comment (joaquinvanschoren) [inline openml/_api_calls.py]: don't you need the authentication flag? GET requests don't require authentication, POST/DELETE requests do.

File: openml/testing.py [hunk 4]
```diff
@@ -65,6 +65,10 @@ def setUp(self):
                 with open(openml.config.config_file, 'w') as fh:
                     fh.write('apikey = %s' % openml.config.apikey)
 
+        # Increase the number of retries to avoid spurios server failures
+        self.connection_n_retries = openml.config.connection_n_retries
+        openml.config.connection_n_retries = 10
```
Comment (joaquinvanschoren) [inline openml/testing.py]: I'm a bit confused about the number of retries being stored in the _default dict, the openml.config file and self.connection_n_retries. So, by default it is 2, then when that doesn't work, it is increased to 10 and stored in the openml.config, and then it stops (it doesn't get higher than 10)?

File: openml/_api_calls.py [hunk 5]
```diff
@@ -102,12 +104,44 @@ def _read_url(url, data=None):
...
+                    raise NotImplementedError()
+                break
+            except (
+                    requests.exceptions.ConnectionError,
+                    requests.exceptions.SSLError,
+            ) as e:
+                if i == n_retries:
+                    raise e
+                else:
+                    time.sleep(0.1 * i)
```
Comment (joaquinvanschoren) [inline openml/_api_calls.py]: nice

File: openml/_api_calls.py (hunk 3 above)
Comment (mfeurer) [reply openml/_api_calls.py]: That is handled internally and does not need to be an argument to the openml-api call API.

File: openml/testing.py (hunk 4 above)
Comment (mfeurer) [reply openml/testing.py]: Nope, it's increased to 10 for testing in every case. It is stored in the class to restore the original value afterwards.

Comment (mfeurer) [reply openml/testing.py]: I agree that it is unfortunate that the value of 2 is twice in the original file, I just made a note to change it later.

**Pull request merged**